### PR TITLE
SI-2704 Handle aggregate signal query aliases

### PR DESCRIPTION
**Pull request by elffjs**
Instead of grouping by time interval and signal name, join `signal` with a temporary table that looks like

|name|agg|
|-|-|
|speed|MIN|
|speed|MAX|
|odometer|AVG|

and group by interval, signal name, and aggregation name. Everything else falls out from that.

### CODE REVIEW

File: schema/signals.tmpl [hunk 1]
```diff
@@ -12,15 +12,15 @@ extend type SignalAggregations {
   {{- else }}
     agg: StringAggregation!
   ):  String
-  {{- end }} @requiresPrivilege(privileges: {{ .Privileges }}) @goField(name: "{{ .GOName }}") @isSignal @hasAggregation
+  {{- end }} @requiresPrivilege(privileges: {{ .Privileges }}) @goField(name: "{{ .GOName }}", forceResolver: true) @isSignal @hasAggregation
```
Comment (elffjs) [inline schema/signals.tmpl:15]: Doing this to avoid having to run code generation for `gqlgen.yml` also.

File: internal/service/ch/ch.go [hunk 2]
```diff
@@ -90,9 +90,17 @@ func (s *Service) GetLatestSignals(ctx context.Context, latestArgs *model.Latest
 // GetAggregatedSignals returns a slice of aggregated signals based on the provided arguments from the ClickHouse database.
 // The signals are sorted by timestamp in ascending order.
 // The timestamp on each signal is for the start of the interval.
-func (s *Service) GetAggregatedSignals(ctx context.Context, aggArgs *model.AggregatedSignalArgs) ([]*vss.Signal, error) {
-	stmt, args := getAggQuery(aggArgs)
-	signals, err := s.getSignals(ctx, stmt, args)
+func (s *Service) GetAggregatedSignals(ctx context.Context, aggArgs *model.AggregatedSignalArgs) ([]*model.AggSignal, error) {
+	if len(aggArgs.FloatArgs) == 0 && len(aggArgs.StringArgs) == 0 {
```
Comment (elffjs) [inline internal/service/ch/ch.go:94]: None of these temporary table things work well when the temporary table is empty, so short circuit here.

File: internal/graph/model/signal_aggs.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+type SignalAggregations struct {
+	Timestamp time.Time `json:"timestamp"`
+
+	ValueNumbers map[AliasKey]float64 `json:"-"`
+	ValueStrings map[AliasKey]string  `json:"-"`
+}
+
+type AggSignal struct {
```
Comment (elffjs) [inline internal/graph/model/signal_aggs.go:17]: Maybe this shouldn't be here.

File: internal/graph/signals_gen.resolvers.tmpl [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+
+	"github.com/DIMO-Network/telemetry-api/internal/graph/model"
+)
+
+{{ range .Signals }}
+// {{ GQLGenResolverName .JSONName }} is the resolver for the {{ .JSONName }}
+{{- if eq .GQLType "Float" }}
+func (r *signalAggregationsResolver) {{ GQLGenResolverName .JSONName }}(ctx context.Context, obj *model.SignalAggregations, agg model.FloatAggregation) (*float64, error) {
+    vn, ok := obj.ValueNumbers[model.AliasKey{Name: "{{ .JSONName }}", Agg: agg.String()}]
```
Comment (elffjs) [inline internal/graph/signals_gen.resolvers.tmpl:15]: This could probably be consolidated more.

File: internal/graph/model/signal_aggs.go (hunk 3 above)
Comment (KevinJoiner) [reply internal/graph/model/signal_aggs.go:17]: It seems appropriate to me. This is just a generic model package for the repo, and this is related to the signal aggregation query.

File: internal/graph/signals_gen.resolvers.tmpl (hunk 4 above)
Comment (KevinJoiner) [reply internal/graph/signals_gen.resolvers.tmpl:15]: Yeah but I don't mind verbose template code. As long as it doesn't decrease the readability.

File: internal/service/ch/queries.go [hunk 5]
```diff
@@ -239,57 +245,79 @@ SELECT
...
 		qm.Where(timestampFrom, aggArgs.FromTS),
 		qm.Where(timestampTo, aggArgs.ToTS),
 		qm.From(vss.TableName),
+		qm.InnerJoin(valueTable),
 		qm.GroupBy(IntervalGroup),
 		qm.GroupBy(vss.NameCol),
+		qm.GroupBy(AggCol),
 		qm.OrderBy(groupAsc),
+		qm.OrderBy(nameAsc), // These last two order-bys are just to make the order standard for tests.
+		qm.OrderBy(aggAsc),
```
Comment (KevinJoiner) [inline internal/service/ch/queries.go:316]: Should we sort the results in the test instead of adding more complexity to the query?

**Review by KevinJoiner — APPROVED** [review]
I really like how this turned out.

**Comment by elffjs** [comment]
What this definitely does not have is deduplication.

File: internal/service/ch/queries.go (hunk 5 above)
Comment (elffjs) [reply internal/service/ch/queries.go:316]: Let's get rid of it.

**Review by Allyson-English — APPROVED** [review]

**Pull request merged**